# Notebook 2: Exploratory Data Analysis (EDA) 


The process conducted in this notebook involves the application of an Exploratory Data Analysis on the ‘candidates’ data migrated to the MySQL database.Basically, Exploratory Data Analysis (EDA) is the process of visually and statistically summarizing, exploring, and understanding the main characteristics, patterns, and relationships within a dataset.

### Importing libraries and modules


`sys.path.append()` allows importing modules from directories that are not in the default Python search path, which is necessary in this case to reuse a module created in the “source/connection/db_utils.py” directory of the project dedicated to the database utilities, like establishing and closing the connection to the database.

In [2]:
import os
import sys
import numpy as np 
import pandas as pd
import matplotlib as plt
import seaborn as sns
sys.path.append(os.path.abspath('../src'))
from connection.db_utils import get_connection, close_connection

### Establishing the database connection

A SQLite engine object is created using the module `get_connection` located in the project´s `/src/connection/db_utils.py` script. Then, a connection object is created by connecting the engine.

In [3]:
engine = get_connection()

Engine created succesfully


### Reading the candidates data from the PosgreSQL database

The pandas library's `read_sql_table` function is employed to retrieve data directly from the PostgresSQL database table named "candidates_raw". This function requires an active database connection, represented by the connection variable, to execute the necessary SQL query. All the data from the "candidates_raw" table is then loaded into a pandas DataFrame, which is assigned to the variable ``df`. This effectively transfers the table's contents into a data structure suitable for analysis and manipulation within Python.

Following the data retrieval, the `df.head(3) method is invoked`. This command displays the first three rows of the DataFrame, providing a concise preview of the data. This allows for immediate verification that the data was successfully extracted from the SQL table and offers a quick look at the structure and initial records of the dataset, facilitating a basic understanding of the data's format and content.

In [4]:
df = pd.read_sql_table("candidates_raw", engine)

df.head(3)

,First Name,Last Name,Email,Application Date,Country,YOE,Seniority,Technology,Code Challenge Score,Technical Interview Score
0,Bernadette,Langworth,leonard91@yahoo.com,26/02/2021,Norway,2,Intern,Data Engineer,3,3
1,Camryn,Reynolds,zelda56@hotmail.com,09/09/2021,Panama,10,Intern,Data Engineer,2,10
2,Larue,Spinka,okey_schultz41@gmail.com,14/04/2020,Belarus,4,Mid-Level,Client Success,10,9


### Data profiling

Data profiling is an invaluable step in the data preparation process. While it doesn’t fully automate data type mapping, it provides crucial insights needed to make informed decisions and create a well-designed database schema in the future.

#### Shape of the data

We get the shape of the data using Pandas `shape` method.

In [5]:
df.shape

(50000, 10)

This means that this dataset has 50000 rows and 10 columns.

#### Data types

We inspect the columns and their data types using the Pandas `info()` method.


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   First Name                 50000 non-null  object
 1   Last Name                  50000 non-null  object
 2   Email                      50000 non-null  object
 3   Application Date           50000 non-null  object
 4   Country                    50000 non-null  object
 5   YOE                        50000 non-null  int64 
 6   Seniority                  50000 non-null  object
 7   Technology                 50000 non-null  object
 8   Code Challenge Score       50000 non-null  int64 
 9   Technical Interview Score  50000 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 3.8+ MB


We can see that the DataFrame contains 3 columns with the "int64" data type and 7 columns with the "object" data type. But "Application Date" should be in a date format.

#### Data cardinality

It helps identify the number of distinct categories in a column, which is essential for analyzing categorical variables.

It measures the "cardinality" of a column, which is the number of unique values. High cardinality means many unique values, while low cardinality means few.

In [9]:
df.nunique()

First Name                    3007
Last Name                      474
Email                        49833
Application Date              1646
Country                        244
YOE                             31
Seniority                        7
Technology                      24
Code Challenge Score            11
Technical Interview Score       11
dtype: int64

* **"First Name 3007":** This means there are 3,007 distinct first names in the "First Name" column.
* **"Last Name 474":** There are 474 unique last names. This indicates less diversity compared to first names, potentially suggesting some last names are repeated more frequently.
* **"Email 49833":** This shows 49,833 unique email addresses. Given that the data likely has 50,000 rows, this suggests that almost every row has a unique email, which is expected, but emails should be unique for each candidates, so these most likely indicates a duplication problem.
* **"Application Date 1646":** There are 1,646 unique application dates. This indicates that applications were submitted on a variety of different days.
* **"Country 244":** There are 244 distinct countries represented in the dataset. This is suspicious there are 195  recognized countries in the world.
* **"YOE 31":** "YOE" likely stands for "Years of Experience." There are 31 unique values, meaning applicants have a range of experience levels.
* **"Seniority 7":** There are 7 unique seniority levels. This suggests a relatively small number of defined seniority categories.
* **"Technology 24":** There are 24 unique technologies listed. This indicates a variety of technical skills among the applicants.
* **"Code Challenge Score 11":** There are 11 unique scores for the code challenge. This likely means the scores are discrete or binned into a limited number of categories.
* **"Technical Interview Score 11":** Similar to the code challenge score, there are 11 unique technical interview scores.

#### Describe

We get quick summary of the dataset using the Pandas `describe()` method. The `describe()` function applies basic statistical computations on the dataset like extreme values, count of data points standard deviation, etc. Any missing value or NaN value is automatically skipped. `describe()` function gives a good picture of the distribution of data.

Note we can also get the description of categorical columns of the dataset if we specify `include ='all'`  in the describe function.




In [8]:
df.describe(include='all') 

,First Name,Last Name,Email,Application Date,Country,YOE,Seniority,Technology,Code Challenge Score,Technical Interview Score
count,50000,50000,50000,50000,50000,50000.000000,50000,50000,50000.000000,50000.000000
unique,3007,474,49833,1646,244,NaN,7,24,NaN,NaN
top,Sarai,Murazik,marianne31@yahoo.com,07/07/2020,Malawi,NaN,Intern,Game Development,NaN,NaN
freq,33,138,3,50,242,NaN,7255,3818,NaN,NaN
mean,NaN,NaN,NaN,NaN,NaN,15.286980,NaN,NaN,4.996400,5.003880
std,NaN,NaN,NaN,NaN,NaN,8.830652,NaN,NaN,3.166896,3.165082
min,NaN,NaN,NaN,NaN,NaN,0.000000,NaN,NaN,0.000000,0.000000
25%,NaN,NaN,NaN,NaN,NaN,8.000000,NaN,NaN,2.000000,2.000000
50%,NaN,NaN,NaN,NaN,NaN,15.000000,NaN,NaN,5.000000,5.000000
75%,NaN,NaN,NaN,NaN,NaN,23.000000,NaN,NaN,8.000000,8.000000


### transformations

In [ ]:
# Use .values to get 'application_date' column as a list
application_date_values = df['application_date'].values.tolist()

# Show the list
print("List of 'application_date' Column")
print(application_date_values)

In [ ]:
# Converting 'application_date' column to datetime

try:
        df['application_date'] = pd.to_datetime(df['application_date'], dayfirst=True)
        print(df['application_date'].iloc[:3])
except (ValueError, TypeError) as e:
        print(f"Error processing date column '{application_date}': {e}")


# Display data types after conversion
print("\nAfter Conversion:")
print(df.info())

In [ ]:
df.duplicated().sum()

In [ ]:
# Find rows with duplicate emails
duplicate_emails = df[df.duplicated(subset=['email'], keep=False)]
duplicate_emails["email"].value_counts()


In [ ]:
print(f"Found {len(duplicate_emails)} rows with duplicate emails.")

In [ ]:
def process_candidate_uniqueness(df):
    """
    Checks for unique candidates (first_name, last_name, email)
    and then checks if those unique candidates have different application dates.
    Prints the number of rows dropped.
    """

    df['unique_candidate'] = True  # Initialize uniqueness flag
    df['different_application_dates'] = False #initialize different dates flag

    # Create candidate key
    df['candidate_key'] = df['first_name'] + '_' + df['last_name']  + '_' + df['email']

    # Find duplicate candidate keys
    duplicate_keys = df[df.duplicated(subset=['candidate_key'], keep=False)]

    # Mark duplicates as False
    df.loc[duplicate_keys.index, 'unique_candidate'] = False

    # Check for different application dates among unique candidates
    unique_candidates = df[df['unique_candidate']]  # Only process unique candidates

    for key, group in unique_candidates.groupby('candidate_key'):
        if len(group['application_date'].unique()) > 1:
            df.loc[group.index, 'different_application_dates'] = True

    #count how many rows will be dropped.
    original_row_count = len(df)
    df_dropped = df[df['unique_candidate'] == False]
    rows_dropped = len(df_dropped)

    #drop the rows.
    df = df[df['unique_candidate'] == True]

    #print the number of rows dropped.
    print(f"Number of rows dropped: {rows_dropped}")

    df.drop('candidate_key', axis=1, inplace=True)
    return df


df = process_candidate_uniqueness(df)

In [ ]:
# Check if the DataFrame is ordered by date
is_ordered = df['application_date'].is_monotonic_increasing

print(is_ordered)

In [ ]:
# Sort the DataFrame by the 'date' column
df_sorted = df.sort_values(by='application_date')

In [ ]:
min_values = df.min()
print(min_values)


In [ ]:
max_values = df.max()
print(max_values)


In [ ]:
df.info()